marlowe-playground-client/src/Wallet.purs

```diff
@@ -0,0 +1,1301 @@
+  mContract <- use _walletLoadedContract
+  case mContract of
+    Nothing -> pure unit -- shouldn't be possible really
+    Just contract -> traverse_ (makePayment contract) newPayments
+  where
+  makePayment contract (Payment party money) = do
+    let
+      owner = fromMaybe mempty (contract ^? (_parties <<< ix party))
+
+      (assets :: Array _) = Map.toUnfoldable $ joinMaps (unwrap money)
+    for assets \(Tuple (Tuple currencySymbol tokenName) amount) -> do
+      let
+        token = Token currencySymbol tokenName
+
+        addMoney Nothing = Just amount
+
+        addMoney (Just v) = Just (v + amount)
+      modifying (_wallets <<< ix owner <<< _assets) (Map.alter addMoney token)
+
+hasHistory :: State -> Boolean
+hasHistory state = fromMaybe false $ state ^? (_loadedMarloweState <<< to NEL.length <<< to ((<) 1))
+
+handleQuery :: forall a m. Query a -> HalogenM State Action ChildSlots Message m (Maybe a)
+handleQuery (LoadContract contractString next) = do
+  case parseContract contractString of
+    Left err -> do
+      assign _loadContractError (Just $ show err)
+      pure $ Just next
+    Right contractTerm -> do
+      case fromTerm contractTerm of
+        Nothing -> pure $ Just next
+        Just contract -> do
+          idx <- use (_contracts <<< to Map.size <<< to (add one))
+          mOpenWallet <- use _openWallet
+          case mOpenWallet of
+            Nothing -> pure $ Just next
+            Just openWallet -> do
+              existingWallets <- use _wallets
+              slot <- use _slot
+              let
+                contractData = gatherContractData contractTerm { parties: mempty, tokens: mempty }
+
+                parties = Set.mapMaybe fromTerm contractData.parties
+
+                tokens = Map.fromFoldable $ Set.map (\token -> Tuple token zero) $ Set.mapMaybe fromTerm contractData.tokens
+
+                newWallets = foldl (walletFromPK tokens) existingWallets parties
+
+                -- A PK party should be owned by the wallet that is that PK but by default it is owned by the primary wallet
+                partiesMap = Map.fromFoldable $ Set.map (mkPartyPair openWallet newWallets) parties
+
+                loadedContract =
+                  { contract
+                  , idx
+                  , owner: openWallet ^. _name
+                  , parties: mempty
+                  , started: false
+                  , marloweState: NEL.singleton (emptyMarloweState slot)
+                  }
+              modifying _wallets (Map.union newWallets)
+              assign _walletLoadedContract (Just loadedContract)
+              modifying _contracts (Map.insert idx loadedContract)
+              assign (_walletLoadedContract <<< _Just <<< _parties) partiesMap
+              modifying (_openWallet <<< _Just <<< _assets) (flip Map.union tokens)
+              modifying _tokens (Set.union (Map.keys tokens))
+              updateContractInState contractString
```

Comment:
I know what you mean but I'm not sure, since I added the `for_`s it looks a bit more like the main clause with some 'error handling' around the edges. What do you think? I could add it to a `where` but I'm not sure if that would just add a bit more misdirection.